funding/manager.go

```diff
@@ -968,7 +968,21 @@ func (f *Manager) stateStep(channel *channeldb.OpenChannel,
 	// fundingLocked was sent to peer, but the channel was not added to the
 	// router graph and the channel announcement was not sent.
 	case fundingLockedSent:
-		err := f.addToRouterGraph(channel, shortChanID)
+		// We must wait until we've received the peer's funding locked
+		// before sending a channel_update according to BOLT#07.
+		received, err := f.receivedFundingLocked(chanID)
+		if err != nil {
+			return fmt.Errorf("failed to check if funding locked "+
+				"was received: %v", err)
+		}
+
+		if !received {
+			// We haven't received FundingLocked, so we'll continue
+			// to the next iteration of the loop.
+			return nil
```

Comment:
Won't this end up causing the loop to spin as it typically assumes it can be advanced all the way to a terminal state.

Reply:
it can quit if the funding manager quits or the message is received - this is by design here as otherwise the update we send might be dropped by the peer. could add a sleep so it's not "tight"

Reply:
>  could add a sleep so it's not "tight"

Yeah I haven't run this yet, but I worry it'll just spin a bunch and soak up CPU. We could add a simple heartbeat check instead.

Reply:
select statement with time after?

Reply:
cant we add some kind of signal between `handleFundingLocked` and here? then we can just wait for the signal

Reply:
We could do that, but then that requires a new map and making sure the code can advance to the next state on a restart if funding locked was already received but channel_update wasn't sent. This way isn't the most efficient, but it is dumb and simple

Reply:
My last comment about restart was a bit off since they should retransmit, but I'd still rather avoid a map

Reply:
Consider that the peer might disconnect _right_ after funding: we'd end up spinning here (without any rate limiting) for hours potentially. I think we want to do a `time.After`based backoff here. We can reuse the `computeNextBackoff` function in the server to handle what the back off should be.

Reply:
I think this will mean we have to store more per-peer state vs sleeping a second per try, unless there is a way to give the funding mgr the last time the peer was online

Reply:
Ah yes, we'll need to keep track of the current backlog _somewhere_. Alternatively we can pass it through the main state transition function (which means we don't need yet-another-map). In the future we might want to delay/rate limit other stuff as well, so I don't think it would be a one time thing. 

If we don't want to distrust other callers, we can make it a functional optional argument. so `stateStep` would gain this optional arg then, for certain internal calls, if they return a `time.Duration` or `time.After` we'd thread it around for the next time `stateStep` would also need to return/thread-thru this value.

Reply:
It doesn't cover all cases, this would still spin if the peer was online but has neglected to send funding_locked

Reply:
I can't find an API that stores the last time a peer was connected. So we would need a map from peer->backlog for storage, even if it's not accessed concurrently. While it does seem better than waiting several seconds per run, it does add more state which I don't like